### DISCUSSION TITLE
Empty metadata in Declarative Base class

### DISCUSSION THREAD

**Discussion by klaydev** [sqlalchemy/sqlalchemy#10305] [Usage Questions] [answered]
Hello, i have this code:
```py
# database.py
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.ext.asyncio import AsyncAttrs


class Base(AsyncAttrs, DeclarativeBase):
    pass
```
```py
# user.py
from sqlalchemy import Integer
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class User(Base):
    __tablename__ = "users"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    ...
```
When i display `Base.metadata.info` or `User.metadata.info` i get `{}`
```py
print(Base.metadata.info) # -> {}
print(User.metadata.info) # -> {}
```

Do I understand correctly that metadata should be inherited by all heirs? Will it be the same for everyone, or is it created for each heir?

**Comment by CaselIT** [comment]
Hi,

I'm not sure what's the question. `info` is a dict for used defined data that's present in most sqlalchemy objects. It's not otherwise used by sqlalchemy

**Reply by klaydev** [reply]
I gave this as an example. If we take tables, then base will have it empty.

**Reply by CaselIT** [reply]
I'm not sure what you mean by empty. What is empty? `info`?

**Reply by klaydev** [reply]
More info:
```py
print(User.metadata.tables) # -> not empty FacadeDict
print(Base.metadata.tables) # -> empty FacadeDict
```

**Reply by klaydev** [reply]
shouldn't the metadata of the base class contain all metadata from all descendant classes?

**Reply by CaselIT** [reply]
Yes it's the same.
Using your example both should be populated with the same data.

Note that you have to import user.py before printing `Base.metadata` for it's table to be registered

**Reply by klaydev** [reply]
It doesn't work.
```py
>>> from models import User
>>> from app.database import Base
>>> Base.metadata.tables
FacadeDict({})
```
If you try to do migration via alembic, it will be empty

**Reply by CaselIT** [reply]
Can you post your complete code? The snippet you posted would work correctly

**Comment by zzzeek** [comment]
>  Do I understand correctly that metadata should be inherited by all heirs? Will it be the same for everyone, or is it created for each heir?

you can answer this question definitively by creating a simple test, as below. Try running this:

```py
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.ext.asyncio import AsyncAttrs


class Base(AsyncAttrs, DeclarativeBase):
    pass


from sqlalchemy import Integer
from sqlalchemy.orm import Mapped, mapped_column



class User(Base):
    __tablename__ = "users"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)


assert Base.metadata is User.metadata
assert list(Base.metadata.tables) == ["users"]
assert list(User.metadata.tables) == ["users"]

print(Base.metadata.tables)
print(User.metadata.tables)
```

in your ``user.py`` file, you can make the same assertions:

```py
# user.py
from sqlalchemy import Integer
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class User(Base):
    __tablename__ = "users"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    ...

assert Base.metadata is User.metadata
assert list(Base.metadata.tables) == ["users"]
assert list(User.metadata.tables) == ["users"]

```


we can see that the Base / User classes have the same collection for `MetaData`.  That's how we confirm what SQLAlchemy the library does.   

now for your actual program / test, we can assume something is going wrong somewhere, which is that either the Base / User being viewed at here do not actually correspond, or that there is something else going on, which could be things like custom metaclasses, attributes explicitly named `metadata`, event hooks, multiple processes running, who knows.    So start from my example above that "works" and work outwards to what your actual application has set up to find where the behavior changes.

**Reply by klaydev** [reply]
1. AssertionError
2. AssertionError
3. Ok

**Reply by klaydev** [reply, marked as answer]
Ouch, i using custom `metadata=` problem is solved!
